src/lightning/fabric/strategies/fsdp.py

```diff
@@ -323,35 +324,31 @@ def setup_optimizer(self, optimizer: Optimizer) -> Optimizer:
     def module_to_device(self, module: Module) -> None:
         pass
 
-    @contextmanager
-    def module_init_context(self, empty_init: Optional[bool] = None) -> Generator[None, None, None]:
-        empty_init_context: Union[torch.device, _EmptyInit, nullcontext]
+    def module_init_context(self, empty_init: Optional[bool] = None) -> ContextManager:
+        stack = ExitStack()
         if _TORCH_GREATER_EQUAL_2_1 and empty_init:
             # Materialization happens in `setup`. When modules get wrapped by FSDP, the sequence of operations is:
             # 1) materialize module 2) call `reset_parameters()` 3) shard the module.
             # These operations are applied to each submodule 'bottom up' in the module hierarchy.
-            empty_init_context = torch.device("meta")
+            stack.enter_context(torch.device("meta"))
         elif _TORCH_GREATER_EQUAL_1_13:
-            empty_init_context = _EmptyInit(enabled=bool(empty_init))
-        else:
-            empty_init_context = nullcontext()
-        with empty_init_context, self.precision.init_context(), self.module_sharded_context():
-            yield
+            stack.enter_context(_EmptyInit(enabled=bool(empty_init)))
+        stack.enter_context(self.precision.init_context())
+        stack.enter_context(self.module_sharded_context())
+        return stack
```

Comment:
qq, could you double check the order here? does entering in this order represent the same order in which the previous declaration was made?

Reply:
It's LIFO:
```python
import contextlib


class A:
    def __enter__(self): print(f"{type(self).__name__} enter")
    def __exit__(self, exc_type, exc_val, exc_tb): print(f"{type(self).__name__} exit")

class B:
    def __enter__(self): print(f"{type(self).__name__} enter")
    def __exit__(self, exc_type, exc_val, exc_tb): print(f"{type(self).__name__} exit")

class Strategy:
    def init_context(self):
        stack = contextlib.ExitStack()
        stack.enter_context(A())
        stack.enter_context(B())
        return stack


def fn():
    strategy = Strategy()
    with strategy.init_context():
        print("Inside")

out = fn()
```

Reply:
Unfortunately `ExitStack` will also graph break. I opened https://github.com/pytorch/pytorch/issues/109309 to ask if it could be supported